File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/I.Ограниченная_очередь/solution.py
import sys


class MyQueueSized:
    def __init__(self, n: int) -> None:
        self.queue: list[int] = [0] * n
        self.head: int = 0
        self.tail: int = 0
        self.capacity: int = n
        self.size_: int = 0

    def push(self, x: int) -> None:
        if self.size_ == self.capacity:
            sys.stdout.write('error\n')
            return
        self.queue[self.tail] = x
        self.tail = (self.tail + 1) % self.capacity
        self.size_ += 1

    def pop(self) -> None:
        if not self.size_:
            sys.stdout.write('None\n')
            return
        sys.stdout.write(str(self.queue[self.head]) + '\n')
        self.queue[self.head] = 0
        self.head = (self.head + 1) % self.capacity
        self.size_ -= 1

    def peek(self) -> None:
        if not self.size_:
            sys.stdout.write('None\n')
            return
        sys.stdout.write(str(self.queue[self.head]) + '\n')

    def size(self) -> None:
        sys.stdout.write(str(self.size_) + '\n')


def solution(t: int) -> None:
    """
    Time Complexity O(N) for building queue and O(1) for each operation
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    queue: MyQueueSized = MyQueueSized(n)
    for _ in range(t):
        command: str = sys.stdin.readline().rstrip()
        if command.startswith('push'):
            queue.push(command.split()[1])
        elif command.startswith('pop'):
            queue.pop()
        elif command.startswith('peek'):
            queue.peek()
        else:
            queue.size()


if __name__ == '__main__':
    t: int = int(sys.stdin.readline().rstrip())
    solution(t)

File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/I.Ограниченная_очередь/solution.cpp
#include <iostream>
#include <vector>
#include <string>

class MySizedQueue {
private:
    std::vector<int> queue;
    int head = 0;
    int tail = 0;
    int size_ = 0;
    int capacity = 0;
public:
    MySizedQueue(int n) {
        queue.resize(n, 0);
        capacity = n;
    };
    void push(int x) {
        if (size_ == capacity) {
            std::cout << "error\n";
            return;
        }
        queue[tail] = x;
        tail = (tail + 1) % capacity;
        ++size_;
    }
    void pop() {
        if (!size_) {
            std::cout << "None\n";
            return;
        }
        std::cout << queue[head] << "\n";
        queue[head] = 0;
        head = (head + 1) % capacity;
        --size_;
    }
    void peek() {
        if (!size_) {
            std::cout << "None\n";
            return;
        }
        std::cout << queue[head] << "\n";
    }
    void size() {
        std::cout << size_ << "\n";
    }
};


void solution(int t) {
    /*
    Time Complexity O(N) for building queue and O(1) for each operation
    Memory Complexity O(N)
    */
    int n;
    std::cin >> n;
    MySizedQueue queue = MySizedQueue(n);
    for (int i = 0; i < t; ++i) {
        std::string command;
        std::cin >> command;
        if (command == "push") {
            int num;
            std::cin >> num;
            queue.push(num);
        } else if (command == "pop") {
            queue.pop();
        } else if (command == "peek") {
            queue.peek();
        } else {
            queue.size();
        }
    }
}

int main() {
    int t;
    std::cin >> t;
    solution(t);
}